Compute volume from cell_area if available

enekomartinmartinez opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

-->

## Description

This PR uses [_try_adding_calculated_cell_area](https://github.com/ESMValGroup/ESMValCore/blob/c8635a843f114374f7bd52aceb3781188d6d35b5/esmvalcore/preprocessor/_area.py#L286) in [calculate_volume](https://github.com/ESMValGroup/ESMValCore/blob/c8635a843f114374f7bd52aceb3781188d6d35b5/esmvalcore/preprocessor/_volume.py#L100), so the volume can be computed from `cell_area` if available in the cube. This opens the possibility of working with `volume_statistics` in irregular grids with `areacello` but not `volcello`.
<!--
    Please describe your changes here, especially focusing on why this pull
    request makes ESMValCore better and what problem it solves.

    Before you start, please read our contribution guidelines: https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html

    Please fill in the GitHub issue that is closed by this pull request,
    e.g. Closes #1903
-->

Closes #2310 

Link to documentation:

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

## Note
I have tried to run tests locally so I can develop a new test but I am having some issues. After installing the environment.yml with conda (`conda env create -f environment.yml`) and ESMValCore in developer mode with pip (`pip install -e .`), I am having the following error message when trying to run the tests:
```
________________________________________________________________ ERROR collecting test session _________________________________________________________________
../../.miniconda/envs/esmvaltool/lib/python3.11/site-packages/pluggy/_hooks.py:501: in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
../../.miniconda/envs/esmvaltool/lib/python3.11/site-packages/pluggy/_manager.py:119: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
../../.miniconda/envs/esmvaltool/lib/python3.11/site-packages/_pytest/python.py:210: in pytest_collect_directory
    if pkginit.is_file():
../../.miniconda/envs/esmvaltool/lib/python3.11/pathlib.py:1267: in is_file
    return S_ISREG(self.stat().st_mode)
../../.miniconda/envs/esmvaltool/lib/python3.11/pathlib.py:1013: in stat
    return os.stat(self, follow_symlinks=follow_symlinks)
E   PermissionError: [Errno 13] Permission denied: '/home/lost+found/__init__.py'
```

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

@schlunma

schlunma commented:
Thanks for the PR!

I am not quite sure what's causing the error here. How are you running the tests? For me, a `cd /your/path/to/ESMValCore && pytest` usually works.

enekomartinmartinez commented:
Thanks @schlunma 

Yes, I am doing a `cd /path/toESMValCore`, and then I have tried several options:

- `pytest tests/unit/preprocessor/_volume/test_volume.py` (this also produces some errors due to raised warnings so I used the default configuration later)
```
======================================== ERRORS =========================================
__________________________________ ERROR collecting . ___________________________________
../../.miniconda/envs/esmvaltool/lib/python3.11/site-packages/pluggy/_hooks.py:501: in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
../../.miniconda/envs/esmvaltool/lib/python3.11/site-packages/pluggy/_manager.py:119: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
../../.miniconda/envs/esmvaltool/lib/python3.11/site-packages/pytest_mypy.py:111: in pytest_collect_file
    return MypyFile.from_parent(parent=parent, fspath=path)
../../.miniconda/envs/esmvaltool/lib/python3.11/site-packages/pytest_mypy.py:123: in from_parent
    return getattr(super(), 'from_parent', cls)(*args, **kwargs)
../../.miniconda/envs/esmvaltool/lib/python3.11/site-packages/_pytest/nodes.py:661: in from_parent
    return super().from_parent(parent=parent, fspath=fspath, path=path, **kw)
../../.miniconda/envs/esmvaltool/lib/python3.11/site-packages/_pytest/nodes.py:275: in from_parent
    return cls._create(parent=parent, **kw)
../../.miniconda/envs/esmvaltool/lib/python3.11/site-packages/_pytest/nodes.py:152: in _create
    return super().__call__(*k, **kw)
../../.miniconda/envs/esmvaltool/lib/python3.11/site-packages/_pytest/nodes.py:616: in __init__
    path = _imply_path(type(self), path, fspath=fspath)
../../.miniconda/envs/esmvaltool/lib/python3.11/site-packages/_pytest/nodes.py:108: in _imply_path
    warnings.warn(
E   pytest.PytestRemovedIn8Warning: The (fspath: py.path.local) argument to MypyFile is deprecated. Please use the (path: pathlib.Path) argument instead.
E   See https://docs.pytest.org/en/latest/deprecations.html#fspath-argument-for-node-constructors-replaced-with-pathlib-path
```

This both options gave me the error from above:
- `pytest -c /dev/null tests/unit/preprocessor/_volume/test_volume.py`

- `pytest -c /dev/null /path/toESMValCore/tests/unit/preprocessor/_volume/test_volume.py`



I don't know where the problem can come from. I have another conda environment with pytest for another library and it works fine. So, it could be related to the specific configuration of ESMValCore tests; maybe I did something wrong? I will try to reinstall the environment and go deeper if it doesn't work when I have more time. I was asking in case the error was familiar to you.

schlunma commented:
Ah, your first error actually looks familiar: https://github.com/ESMValGroup/ESMValCore/issues/2314

This has been fixed, but requires setting up a new environemt. Sorry, this is really unlucky timing.

enekomartinmartinez commented:
Thanks, I will then create the new environment and try to add new tests

enekomartinmartinez commented:
Hi @schlunma 

I am still working on this, here are some updates with the last commit:

1. Now `calculate_volume` is more flexible, previous implementation needed to have always z in the second dimensions `z_dim == 1`. I have made that more generic so this preprocessor can run in models where z is in order dimension (e.g. a cube after a `climate_statistics` with no time dimension).
2. It also works with non-regular data if `areacello` available

I have tested running volume_statistics over regular data with no areacello and with irregular data with areacello and seems to run fine.

I still have to include the test.

enekomartinmartinez commented:
Fixed tests and added new ones, I made some new error messages for the following cases:

- z-axis bounds not 2 in last dim -> cannot compute thickness
- xyz_dims > xy_dims +1 -> z depends on other non-spatial dimensions (like time) this would provoke aggregating the cube also in that dimension (this is an adaptation of a previous error)

Still missing test when cell_measure is available

enekomartinmartinez commented:
Hi @schlunma,

This is ready to merge by my side

schlunma reviewed (changes requested):
Thanks for these change @enekomartinmartinez, looks really good already!

I have a couple of comments directly on the code.

In addition, it would be great if you could make sure that `areacello` is loaded automatically if no supplementary is specified (this will not raise an error if that is not found). You can do that by adding it to the decorator [here](https://github.com/ESMValGroup/ESMValCore/blob/bb7866e6a136f6e011bc6419c4e3ed9eec25721e/esmvalcore/preprocessor/_volume.py#L182-L185). Also, please make sure the documentation is up-to-date (i.e., [here](https://github.com/ESMValGroup/ESMValCore/blob/bb7866e6a136f6e011bc6419c4e3ed9eec25721e/doc/recipe/preprocessor.rst#L2136-L2140) and [here](https://github.com/ESMValGroup/ESMValCore/blob/bb7866e6a136f6e011bc6419c4e3ed9eec25721e/doc/recipe/preprocessor.rst#L310)).

Thanks!!

enekomartinmartinez commented:
Thank you @schlunma for so many comments and suggestions, I will be accepting the changes and answering them one by one!

enekomartinmartinez commented on tests/unit/preprocessor/_volume/test_volume.py:
```diff
@@ -443,49 +504,100 @@ def test_volume_statistics_wrong_operator_fail(self):
         self.grid_4d_z.add_aux_coord(new_z_coord, (1, 2))
-        with self.assertRaises(ValueError) as err:
-            volume_statistics(self.grid_4d_z, 'mean')
-        self.assertIn(
```
Yes, in fact this tests existed before this PR, I only changed it how it works (to make it more simple) as I create a irregular data for the test with arecello available. Please, check original here
https://github.com/ESMValGroup/ESMValCore/blob/bb7866e6a136f6e011bc6419c4e3ed9eec25721e/tests/unit/preprocessor/_volume/test_volume.py#L445-L452

New version here
https://github.com/enekomartinmartinez/ESMValCore/blob/bb4b4ccac98a52b2a1f9c9bbc175da3f9cb2caf5/tests/unit/preprocessor/_volume/test_volume.py#L506

enekomartinmartinez commented on esmvalcore/preprocessor/_volume.py:
```diff
@@ -227,6 +244,20 @@ def volume_statistics(
+    # assert z axis only uses 1 dimension more than x, y axis
+    xy_dims = tuple({*cube.coord_dims(y_axis), *cube.coord_dims(x_axis)})
+    xyz_dims = tuple({*cube.coord_dims(z_axis), *xy_dims})
+    if len(xyz_dims) > len(xy_dims) + 1:
```
I tried to run this test here with the new version:
https://github.com/ESMValGroup/ESMValCore/blob/bb7866e6a136f6e011bc6419c4e3ed9eec25721e/tests/unit/preprocessor/_volume/test_volume.py#L454-L462

where the z-coord (output from `cube.coord(axis='z')`) has 4D. In that case, z-coord depends on (0, 1, 2, 3), so the dimensions correspond to time, z, y, x. When aggregating the data as z-coord was (0, 1, 2, 3), the dimension corresponding to time was also aggregated, so the average was computed also on time. I don't know if this could be an issue with real data. However, as there was already a test that could be problematic, I decided to include this error. 

Then, the idea is to prevent aggregations over time dimension if the z-coord depends on time. A better solution would be to use a dim-coord that only represents the z-axis, but I am unsure how that is presented in real data. In that case, we should properly differentiate between the z-dim used to do the aggregation, which should be 1D, and z-coord, which could have more dimensions and is used to compute the thickness. 

But again, I am not sure which is the specific case of all the datasets, so I am not sure if this is a real problem and if it is how to implement the aggregation without including bugs.

enekomartinmartinez replied:
Note that I kept the original test, but I changed its name (just to make it more specific about the error)
https://github.com/enekomartinmartinez/ESMValCore/blob/bb4b4ccac98a52b2a1f9c9bbc175da3f9cb2caf5/tests/unit/preprocessor/_volume/test_volume.py#L541

schlunma replied:
Ah, I see, that makes sense. Thanks for the explanation.

enekomartinmartinez commented on esmvalcore/preprocessor/_volume.py:
```diff
@@ -227,6 +244,20 @@ def volume_statistics(
+            f"X and Y axis coordinates depend on {xy_dims} dimensions, "
+            f"while X, Y, and Z axis depends on {xyz_dims} dimensions. "
+            "This may indicate Z axis depending on other dimension than"
+            "space that could provoke invalid aggregation...")
```
I explained above. Let me know your thoughts about that. If necessary, I could rewrite this error message to make it more easy to understand.

enekomartinmartinez commented on esmvalcore/preprocessor/_volume.py:
```diff
@@ -125,33 +124,46 @@ def calculate_volume(cube: Cube) -> da.core.Array:
+    if not has_cell_measure:
+        cube.remove_cell_measure('cell_area')
+
+    # Get slices to make area and thickness match with the same number
+    # of dimensions as the cube
+    z_slice = tuple(
+        slice(None) if i in z_dim else None
+        for i in range(cube.ndim)
+    )
+    area_slice = tuple(
+        slice(None) if i in area_dim else None
+        for i in range(cube.ndim)
+    )
+
+    # Calculate grid cell volume as area * thickness
+    grid_volume = area.lazy_data()[area_slice] * thickness[z_slice]
```
Thanks, it's much simpler. The only question I have is if this iris function keeps the data lazy. Checking the iris source code, it seems that it doesn't, but I haven't checked. 

In any case, maybe it is better to use this function and ask iris to make the changes?

schlunma replied:
Yes, since [Iris 3.7](https://scitools-iris.readthedocs.io/en/latest/whatsnew/3.7.html#features), this function maintains lazy data. If it doesn't do that locally for you, you might need to update your installation.

enekomartinmartinez replied:
My bad, I didn't check the documentation for the latest stable version! Thanks for that :)

enekomartinmartinez commented on esmvalcore/preprocessor/_volume.py:
```diff
@@ -227,6 +244,20 @@ def volume_statistics(
+        raise ValueError(
+            f"X and Y axis coordinates depend on {xy_dims} dimensions, "
+            f"while X, Y, and Z axis depends on {xyz_dims} dimensions. "
+            "This may indicate Z axis depending on other dimension than"
```
Nice catch, I corrected this locally to correct it also on tests at the same time. Force updated with the other suggestions

enekomartinmartinez commented:
> Thanks for these change @enekomartinmartinez, looks really good already!
> 
> I have a couple of comments directly on the code.
> 
> In addition, it would be great if you could make sure that `areacello` is loaded automatically if no supplementary is specified (this will not raise an error if that is not found). You can do that by adding it to the decorator [here](https://github.com/ESMValGroup/ESMValCore/blob/bb7866e6a136f6e011bc6419c4e3ed9eec25721e/esmvalcore/preprocessor/_volume.py#L182-L185). Also, please make sure the documentation is up-to-date (i.e., [here](https://github.com/ESMValGroup/ESMValCore/blob/bb7866e6a136f6e011bc6419c4e3ed9eec25721e/doc/recipe/preprocessor.rst#L2136-L2140) and [here](https://github.com/ESMValGroup/ESMValCore/blob/bb7866e6a136f6e011bc6419c4e3ed9eec25721e/doc/recipe/preprocessor.rst#L310)).
> 
> Thanks!!

I am working in this other suggestions, I still have few questions. Should I register as well `arecella` in the case `volume_statistics` is used with atmospheric variables?

Shouldn't we check in `calculate_volume` that z-coord is in meters, so we ensure always that the returned volume is in cubic meters?

schlunma commented:
> Should I register as well `arecella` in the case `volume_statistics` is used with atmospheric variables?

Good point. The function has clearly been designed for ocean data, but it would of course be nice to also support atmospheric variables. However, to properly support this we would probably need to consider pressure levels in addition to altitudes. Let's do that in a separate PR.

> Shouldn't we check in `calculate_volume` that z-coord is in meters, so we ensure always that the returned volume is in cubic meters?

Also good point! You could also return the multiplied units of the area and z dimension in `calculate_volume` and then use that as units for the cell measure. Then it would also work for z-coords with different units.

enekomartinmartinez commented:
> Good point. The function has clearly been designed for ocean data, but it would of course be nice to also support atmospheric variables. However, to properly support this we would probably need to consider pressure levels in addition to altitudes. Let's do that in a separate PR.

I cannot currently work on further development. I don't know if this is a requirement but there may be people already using the preprocessor with atmospheric data as it may not fail and use the thickness of pressure_levels automatically. Should this be considered risky?

> Also good point! You could also return the multiplied units of the area and z dimension in `calculate_volume` and then use that as units for the cell measure. Then it would also work for z-coords with different units.

My question is should we allow using things like `area * thickness in sigma space` as a correct volume measure? Wouldn't the weights (kg/m) be incorrect and produce an unrealistic product?

schlunma commented:
> I cannot currently work on further development. I don't know if this is a requirement but there may be people already using the preprocessor with atmospheric data as it may not fail and use the thickness of pressure_levels automatically. Should this be considered risky?

I am not saying you need to do this 😉 If you think adding `areacella` is useful, I am happy with it. In this case it might be a good idea to add the `['volcello', 'areacello', 'areacella']` combination to [this](https://github.com/ESMValGroup/ESMValCore/blob/unstructured_linear_regrid/esmvalcore/_recipe/to_datasets.py#L192) function, so that for ocean variables we get  `['volcello', 'areacello']`, and for others `['areacella']`. It would probably also be a good idea to a add a note to the preprocessor that states that this preprocessor has been designed for oceanic variables, but it might be applicable to atmospheric data as well.
 
> My question is should we allow using things like `area * thickness in sigma space` as a correct volume measure? Wouldn't the weights (kg/m) be incorrect and produce an unrealistic product?

Ah, I didn't consider sigma space. If the units can be converted to `m` then it shouldn't be a problem (the iris function will automatically adapt the target units if necessary), but if that's not the case this is problematic. So a test if the units are convertible to `m` would be nice indeed!

enekomartinmartinez commented:
Hi @schlunma 

I have added a conversion of depth units to meters in `calculate_volume` and a test for a cube in sigma space with no volcello given, where the conversion fails. I have documented it in the same function, `volume_statistics` and in `doc/recipe/preprocessor.rst`. I have also added a note in  `volume_statistics`  for the atmospheric case.

About adding `areacella`, as you prefer, maybe we can leave it without it. Then, in case anyone wants to include a conversion for pressure levels to meters, include it.

schlunma commented on esmvalcore/preprocessor/_volume.py:
```diff
@@ -125,33 +126,52 @@ def calculate_volume(cube: Cube) -> da.core.Array:
-            cube.summary(shorten=True),
-        )
-        raise
+    # Convert units back
+    depth.convert_units(original_units)
```
Why are you converting them back? If you want to have your output in `m3` (which the docstring says, you need to leave them in `m`)

enekomartinmartinez replied:
It's done after computing the thickness, so that value is kept in `m`. But it is true that would be much cleaner copying the depth, thankyou :)

schlunma commented on esmvalcore/preprocessor/_volume.py:
```diff
@@ -125,33 +126,52 @@ def calculate_volume(cube: Cube) -> da.core.Array:
     # Load depth field and figure out which dim is which
     depth = cube.coord(axis='z')
-    z_dim = cube.coord_dims(depth)[0]
+    z_dim = cube.coord_dims(depth)
```
```suggestion
    z_dim = cube.coord_dims(depth)
    depth = depth.copy()
```

To avoid overwriting the input cube

schlunma commented on esmvalcore/preprocessor/_volume.py:
```diff
@@ -125,33 +126,52 @@ def calculate_volume(cube: Cube) -> da.core.Array:
+    # Get or calculate the horizontal areas of the cube
+    has_cell_measure = bool(cube.cell_measures('cell_area'))
+    _try_adding_calculated_cell_area(cube)
+    area = cube.cell_measure('cell_area')
```
```suggestion
    area = cube.cell_measure('cell_area').copy()
```

To avoid overwriting the input cube

schlunma reviewed (commented):
Thanks @enekomartinmartinez for these changes! I have one remaining comment about not changing the input cube in-place, then I will approve. Thanks for your patience!!

schlunma reviewed (approved):
Great! Thanks for your contribution 🚀
